File: src/extractors.js
import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";

const textExtensions = new Set([".md", ".txt", ".csv", ".tsv", ".json", ".html", ".htm"]);
const plannedBinaryExtensions = new Set([".pdf", ".docx", ".pptx", ".xlsx", ".png", ".jpg", ".jpeg", ".tiff", ".bmp", ".webp"]);
const parserScript = path.resolve("tools", "parse_document.py");

export function supportedExtensions() {
  return [...textExtensions, ...plannedBinaryExtensions].sort();
}

export function parserStatus() {
  const python = resolvePython();
  if (!python) {
    return {
      python: null,
      engines: {},
      ready: false,
      message: "No Python executable found. Set MEM_PYTHON to enable Docling/MarkItDown parsers."
    };
  }
  const probe = spawnSync(python, [parserScript, "--probe"], {
    cwd: process.cwd(),
    encoding: "utf8",
    windowsHide: true
  });
  if (probe.status !== 0) {
    return {
      python,
      engines: {},
      ready: false,
      message: probe.stderr?.trim() || probe.stdout?.trim() || "Parser probe failed."
    };
  }
  return JSON.parse(probe.stdout);
}

export function extractDocument(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const title = path.basename(filePath);
  if (textExtensions.has(ext)) {
    const raw = fs.readFileSync(filePath, "utf8");
    const markdown = normalizeTextToMarkdown(raw, ext, title);
    return {
      parser: "native-text",
      title,
      type: ext.slice(1) || "text",
      markdown,
      structured: {
        title,
        type: ext.slice(1) || "text",
        sections: [{ kind: "document", heading: title, text: markdown }]
      }
    };
  }

  if (plannedBinaryExtensions.has(ext)) {
    return extractWithPython(filePath, title, ext);
  }

  throw new Error(`Unsupported file extension: ${ext || "(none)"}`);
}

function extractWithPython(filePath, title, ext) {
  const python = resolvePython();
  if (!python) {
    throw new Error("Python parser unavailable. Install Python or set MEM_PYTHON to a Python executable with Docling/MarkItDown.");
  }

  const result = spawnSync(python, [parserScript, filePath], {
    cwd: process.cwd(),
    encoding: "utf8",
    maxBuffer: 64 * 1024 * 1024,
    windowsHide: true
  });
  if (result.status !== 0) {
    throw new Error((result.stderr || result.stdout || "").trim() || `Parser failed for ${ext}`);
  }

  const parsed = JSON.parse(result.stdout);
  const markdown = parsed.markdown?.trim();
  if (!markdown) throw new Error(`Parser produced no Markdown for ${title}`);
  return {
    parser: parsed.parser,
    title: parsed.title || title,
    type: parsed.type || ext.slice(1),
    markdown,
    structured: parsed.structured || {
      title: parsed.title || title,
      type: parsed.type || ext.slice(1),
      sections: [{ kind: "document", heading: parsed.title || title, text: markdown }]
    }
  };
}

function resolvePython() {
  const candidates = [
    process.env.MEM_PYTHON,
    "C:\\Users\\abelj\\.cache\\codex-runtimes\\codex-primary-runtime\\dependencies\\python\\python.exe",
    "python",
    "py"
  ].filter(Boolean);

  for (const candidate of candidates) {
    const check = spawnSync(candidate, ["--version"], {
      encoding: "utf8",
      windowsHide: true
    });
    if (check.status === 0) return candidate;
  }
  return null;
}

function normalizeTextToMarkdown(raw, ext, title) {
  const trimmed = raw.replace(/\r\n/g, "\n").trim();
  if (ext === ".md") return trimmed;
  if (ext === ".csv" || ext === ".tsv") {
    const delimiter = ext === ".tsv" ? "\t" : ",";
    const lines = trimmed.split("\n").filter(Boolean);
    if (lines.length === 0) return `# ${title}\n`;
    const cells = lines.map((line) => line.split(delimiter).map((cell) => cell.trim()));
    const header = cells[0];
    const separator = header.map(() => "---");
    const rows = cells.slice(1);
    return [
      `# ${title}`,
      "",
      `| ${header.join(" | ")} |`,
      `| ${separator.join(" | ")} |`,
      ...rows.map((row) => `| ${row.join(" | ")} |`)
    ].join("\n");
  }
  if (ext === ".html" || ext === ".htm") {
    return `# ${title}\n\n${trimmed
      .replace(/<script[\s\S]*?<\/script>/gi, "")
      .replace(/<style[\s\S]*?<\/style>/gi, "")
      .replace(/<\/(h1|h2|h3|p|li|tr)>/gi, "\n")
      .replace(/<[^>]+>/g, " ")
      .replace(/[ \t]+/g, " ")
      .trim()}`;
  }
  return `# ${title}\n\n${trimmed}`;
}

File: rust/extractors.rs
use anyhow::{Result, anyhow};
use serde::{Deserialize, Serialize};
use std::{fs, path::Path, process::Command};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StructuredDocument {
    pub title: String,
    #[serde(rename = "type")]
    pub doc_type: String,
    #[serde(default)]
    pub sections: Vec<Section>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Section {
    pub kind: Option<String>,
    pub heading: Option<String>,
    pub text: Option<String>,
    pub page: Option<i64>,
    pub slide: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtractedDocument {
    pub parser: String,
    pub title: String,
    #[serde(rename = "type")]
    pub doc_type: String,
    pub markdown: String,
    pub structured: StructuredDocument,
}

pub fn supported_extension(path: &Path) -> bool {
    matches!(
        path.extension().and_then(|e| e.to_str()).map(|e| e.to_lowercase()).as_deref(),
        Some("md" | "txt" | "csv" | "tsv" | "json" | "html" | "htm" | "pdf" | "docx" | "pptx" | "xlsx" | "png" | "jpg" | "jpeg" | "tiff" | "bmp" | "webp")
    )
}

pub fn extract_document(path: &Path) -> Result<ExtractedDocument> {
    let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("").to_lowercase();
    let title = path.file_name().and_then(|n| n.to_str()).unwrap_or("document").to_string();
    match ext.as_str() {
        "md" | "txt" | "csv" | "tsv" | "json" | "html" | "htm" => extract_text(path, &title, &ext),
        _ => extract_with_python(path),
    }
}

pub fn parser_status() -> serde_json::Value {
    let Some(python) = resolve_python() else {
        return serde_json::json!({ "python": null, "engines": {}, "ready": false });
    };
    let output = Command::new(&python)
        .arg("tools/parse_document.py")
        .arg("--probe")
        .output();
    match output {
        Ok(output) if output.status.success() => serde_json::from_slice(&output.stdout).unwrap_or_else(|_| serde_json::json!({ "python": python, "ready": false })),
        Ok(output) => serde_json::json!({ "python": python, "ready": false, "message": String::from_utf8_lossy(&output.stderr) }),
        Err(error) => serde_json::json!({ "python": python, "ready": false, "message": error.to_string() }),
    }
}

fn extract_text(path: &Path, title: &str, ext: &str) -> Result<ExtractedDocument> {
    let raw = fs::read_to_string(path)?;
    let markdown = normalize_text(&raw, ext, title);
    Ok(ExtractedDocument {
        parser: "native-rust-text".to_string(),
        title: title.to_string(),
        doc_type: ext.to_string(),
        structured: StructuredDocument {
            title: title.to_string(),
            doc_type: ext.to_string(),
            sections: vec![Section {
                kind: Some("document".to_string()),
                heading: Some(title.to_string()),
                text: Some(markdown.clone()),
                page: None,
                slide: None,
            }],
        },
        markdown,
    })
}

fn extract_with_python(path: &Path) -> Result<ExtractedDocument> {
    let python = resolve_python().ok_or_else(|| anyhow!("Python parser unavailable. Set MEM_PYTHON."))?;
    let output = Command::new(python)
        .arg("tools/parse_document.py")
        .arg(path)
        .output()?;
    if !output.status.success() {
        return Err(anyhow!("{}", String::from_utf8_lossy(&output.stderr).trim()));
    }
    Ok(serde_json::from_slice(&output.stdout)?)
}

fn resolve_python() -> Option<String> {
    let candidates = [
        std::env::var("MEM_PYTHON").ok(),
        Some("python3".to_string()),
        Some("python".to_string()),
        Some("py".to_string()),
    ];
    candidates.into_iter().flatten().find(|candidate| {
        Command::new(candidate)
            .arg("--version")
            .output()
            .map(|output| output.status.success())
            .unwrap_or(false)
    })
}

fn normalize_text(raw: &str, ext: &str, title: &str) -> String {
    let trimmed = raw.replace("\r\n", "\n").trim().to_string();
    if ext == "md" {
        return trimmed;
    }
    format!("# {title}\n\n{trimmed}")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parser_status_does_not_panic() {
        let status = parser_status();
        assert!(status.get("ready").is_some());
    }
}
